infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -712,39 +718,60 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
     final Map<String, InternalMachineConfig> machineConfigs = environment.getMachines();
     final String workspaceId = getContext().getIdentity().getWorkspaceId();
     LOG.debug("Begin pods creation for workspace '{}'", workspaceId);
-    for (Pod toCreate : environment.getPods().values()) {
-      startTracingContainersStartup(toCreate);
+    for (Pod toCreate : environment.getPodsCopy().values()) {
+      startTracingContainersStartup(toCreate.getSpec(), toCreate.getMetadata());
+      ObjectMeta toCreateMeta = toCreate.getMetadata();
       final Pod createdPod = namespace.deployments().deploy(toCreate);
-      LOG.debug(
-          "Creating pod '{}' in workspace '{}'", toCreate.getMetadata().getName(), workspaceId);
-      final ObjectMeta podMetadata = createdPod.getMetadata();
-      for (Container container : createdPod.getSpec().getContainers()) {
-        String machineName = Names.machineName(toCreate, container);
-
-        LOG.debug("Creating machine '{}' in workspace '{}'", machineName, workspaceId);
-        machines.put(
-            getContext().getIdentity(),
-            new KubernetesMachineImpl(
-                workspaceId,
-                machineName,
-                podMetadata.getName(),
-                container.getName(),
-                MachineStatus.STARTING,
-                machineConfigs.get(machineName).getAttributes(),
-                serverResolver.resolve(machineName)));
-        eventPublisher.sendStartingEvent(machineName, getContext().getIdentity());
-      }
+      LOG.debug("Creating pod '{}' in workspace '{}'", toCreateMeta.getName(), workspaceId);
+      storeStartingMachine(createdPod, toCreateMeta, machineConfigs, serverResolver);
+    }
+    for (Deployment toCreate : environment.getDeploymentsCopy().values()) {
+      PodTemplateSpec template = toCreate.getSpec().getTemplate();
```

Comment:
Most of the code added is a refactor to avoid code duplication (since we do basically the same thing for pods and then for deployments.

I'll add tests covering deployments.

Reply:
Added tests for creating workspace with environment that includes only a deployment and also a deployment and pod.